### ISSUE TITLE
Champion "Support for == and != on tuple types"

### ISSUE THREAD

**Issue by gafter** [dotnet/csharplang#190]
- [ ] Proposal added
- [ ] Discussed in LDM
- [ ] Decision in LDM
- [ ] Finalized (done, rejected, inactive)
- [ ] Spec'ed

See also https://github.com/dotnet/roslyn/issues/13155

**Comment by Thaina** [comment]
Should just support extension operator on IEquatable<T>

**Comment by gafter** [comment]
@Thaina No, `==` and `.Equals` are different things. Try `double.NaN` for example. Also, the `==` operator would convert the values in a tuple, elementwise, to a common type for comparison purposes, while `.Equals` requires that they be the same type.

**Comment by DavidArno** [comment]
Retracting my up-vote for this proposal. As I read "No, `==` and `.Equals` are different things", I do not just witness the principle of least astonishment being violated, it's being smashed into tiny pieces and ground into the dirt.

Yes, for historical reasons, we have ended up with multiple ways of testing for equality and this is one of those "we are stuck with it" things that can't be undone. But perpetuating this messy situation by having new types like `ValueTuple<...>` implement different behaviour for `==` and `.Equals` is pure madness IMO.

We would be better off with the team just leaving this topic well alone and implementing the "extension everything" proposal instead. Then 3rd party libraries can implement `==` and `.Equals` as the same thing and the great disturbance in the force will fade away...

**Comment by ig-sinicyn** [comment]
@DavidArno 

> We would be better off with the team just leaving this topic well alone and implementing the "extension everything" proposal instead. Then 3rd party libraries can implement == and .Equals as the same thing and the great disturbance in the force will fade away...

so, `double.Nan == double.Nan` should be false but
`(double.Nan, double.Nan) == (double.Nan, double.Nan)` 
will be true, false, NaN or a tiny pink elephant depending on a developer's preferences? Pure evil 👍

**Comment by HaloFour** [comment]
Tuples being just a loosely bound sequence of values I think that it makes more sense to be able to compare those values irrespective of the container.

```cs
(int, int) tuple1 = ...;
(int, int) tuple2 = ...;

if (tuple1 == tuple2) { ... }
// the same as
if ((tuple1.Item1 == tuple2.Item1 && tuple1.Item2 == tuple2.Item2)) { ... }
```

**Comment by DavidArno** [comment]
@ig-sinicyn,

The following expression, `result` should be true:
````cs
var x = double.Nan;
var y = double.Nan;
var result = (x == y) == x.Equals(y);
````
But it's `false` as `double` implements `==` and `.Equals` differently. 

You appear to be proposing that, just because the CLR and language teams couldn't agree on whether `NaN` is the same as `NaN`, then tuples equality should perpetuate this nonsense. having `(double.Nan, double.Nan) == (double.Nan, double.Nan)` would be a tiny price to pay for equality sanity in tuples.

**Comment by DavidArno** [comment]
@HaloFour,

````cs
struct S { }

(S, S) tuple1 = ...;
(S, S) tuple2 = ...;

if ((tuple1.Item1 == tuple2.Item1 && tuple1.Item2 == tuple2.Item2)) // oh dear: compiler error
````

`ValueTuple<T1, T2>` cannot implement `==` as the above as it's not constrained to `T1` and `T2` being classes.

**Comment by HaloFour** [comment]
@DavidArno 

Yes, and I have no problem with that.  I don't think that the fact that the two values are contained within a tuple should matter.

**Comment by orthoxerox** [comment]
@DavidArno Did you know that comparison operators and `CompareTo` produce different results on `double`s as well? That's not something that .Net invented either. Comparison and total order operations are defined differently on doubles by IEEE. `Equals` matches the behavior of `CompareTo`.

**Comment by DavidArno** [comment]
@orthoxerox, @HaloFour, @ig-sinicyn,

So you have all done a lot of "no, no, can't be `.Equals()` and @gafter has ruled out the `IEquatable<T>/EqualityComparer` route and `==` on `T1` etc can't be used. So what solution is left?

**Comment by HaloFour** [comment]
@DavidArno 

The solution is that the developer has to use `Equals` manually rather than relying on `==` or `!=`, which is the same answer as when tuples aren't involved.

**Comment by orthoxerox** [comment]
@DavidArno the compiler has to rewrite the `left == right` expression as `(left.Item1 == right.Item1 && left.Item2 == right.Item2)`.

**Comment by DavidArno** [comment]
@orthoxerox,

The following would have to be a compiler error therefore, as `==` in tuples would be a compiler trick, rather than a static operator on the type and so it cannot determine the type at compile-time:
````cs
bool F<T1,T2>((T1 x, T2 y)) => x == y;
````

**Comment by orthoxerox** [comment]
@DavidArno it's an error now and I don't see how it is related to equality of tuples.

**Comment by DavidArno** [comment]
@orthoxerox,

Apologies: the code was wrong. I've updated it to what I meant to say.

**Comment by HaloFour** [comment]
@DavidArno 

So, how about the compiler first attempts to resolve `==` operators defined on `ValueTuple<...>` via extensions and if one isn't found it instead falls back to evaluating the elements individually?

I don't know that I like that, though.  If you needed the container of those two values to matter, enough to define their equality, then you probably shouldn't be using tuples and you should define a proper type instead.

**Comment by orthoxerox** [comment]
@DavidArno Yes, it looks like it will have to be a compiler error, just like this function is right now:

```c#
bool F<T1,T2>(T1 x1, T2 x2, T1 y1, T2 y2) => x1 == y1 && x2 == y2;
```

P.S. I've just spent an hour enumerating all the equality mechanisms of the CLR and I've yet to cover the interfaces.

**Comment by CyrusNajmabadi** [comment]
> You appear to be proposing that, just because the CLR and language teams couldn't agree on whether NaN is the same as NaN

That's not what's happening.

==  implement IEEE754 semantics.    .Equals impelments .Net equals semantics.  IEEE requires that == not be reflexive for NaN.  However, .Equals is required to be reflexive.  i.e. object.Equals states that "a.Equals(a)" must be true for all instances.

If we broke IEEE semantics that would be a problem for tons of customers.  If we broke .net semantics that would be a problem for tons of .net use cases (for example, you could not use a double effectively as a key inside a hashtable).  This approach gives both sets of customers the semantics they require.

**Comment by CyrusNajmabadi** [comment]
> ValueTuple<...> implement different behaviour for == and .Equals is pure madness IMO.

No, it really isn't.  For example, today, if i have:

```C#
byte b = 1;
int i = 1;

Console.WriteLine(b == i);
Console.WriteLine(b.Equals(i));
```

I will print "True, False".

```==``` and ```.Equals``` are already *very different*.

The problem compounds with tuples.  Tuples just aggregate data.  If you aggregate data that uses to compare the same with == but now compares differently (or vice versa), then tuples no longer act properly a composition concept.

**Comment by CyrusNajmabadi** [comment]
Using hte above example, if we do not support == on tuples then you can get the following badness:

```c#
byte b;
string s;

(byte, string) t1 = (b, s);
(int, string) t2 = (b, s);

Console.WriteLine(t1.Item1 == t2.Item1 && t2.Item2 == t2.Item2);
Console.WriteLine(t1 == t2); // doesn't currently compile.  But we would like it to.  Can't be implemented through .Equals
Console.WriteLine(t1.Equals(t2));
```

The top and bottom lines will print 'true', and 'false'.  I strongly believe the second line shoudl print 'true' and that it should just be exactly transformed as:

```c#
t1 == t2 
// becomes
t1.Item1 == t2.Item1 && ... && t1.ItemN == t2.ItemN
```

This, of course, then means that nested tuples work properly (something that definitely does not work with .Equals).

**Comment by DavidArno** [comment]
Ha, ha, thanks @CyrusNajmabadi, the idea of someone trying to use a double as a key in a hashtable is going to keep me chuckling all day! 😁 Even JavaScript, where it'll let black equal white every other Thursday except on leap years doesn't treat NaN as equal. The idea that the .NET team's dogma was so strong that they'd break an international standard just to enable doubles to be used as Dictionary keys is mind-boggling.

Anyway, that aside, you didn't address the important point from above. If a value tuples' types are structs, they may not implement `==` themselves. So how would you propose this feature work?

**Comment by CyrusNajmabadi** [comment]
> the idea of someone trying to use a double as a key in a hashtable is going to keep me chuckling all day!

Using a double as a key is completely fine to do.  

> The idea that the .NET team's dogma was so strong that they'd break an international standard just to enable doubles to be used as Dictionary keys is mind-boggling.

It's interesting that you complain about things being inconsistent, but then complain when .net insisted that .Equals behave consistently :)

> Anyway, that aside, you didn't address the important point from above. If a value tuples' types are structs, they may not implement == themselves. So how would you propose this feature work?

The exact way that == works for types that don't have an available == operator.  I'm stating that == for tuples is defined as:

```c#
(x1, ..., xn) == (y1, ..., yn)   <==>  x1 == y1 && ... && xn == yn
```

== on the tuple will work in precisely all the circumstances that == works for the constituent elements.

> just to enable doubles to be used as Dictionary keys is mind-boggling.

I never said that.  Please do not put words in my mouth.  I simply gave an example of how the .Equals behavior is desirable and consistent.  I did not say that it was done "just" for this reason.

--

Also, to be clear, you never addressed the other points i made.  For example, that == and .Equals are not equivalent today for *ordinary numeric types* other than double.  Today == will return true when .Equals does not.  == follows the rules of C# while .Equals follows the rules of .Net.   Making it so that we have a type in C# which composes other C# types, but which does not compose == appropriately just leads to the composition not actually happening fully.

**Comment by CyrusNajmabadi** [comment]
Note: having .Equals be symmetric is nice for lots of reasons.  After all, while it would be very weird to have "a.Equals(a)" be false, it would be even more confusing and difficult to use .net effectively if doubles didn't support this concept.  After all, the following could then be false:

```c#
list.Add(d);
Console.WriteLine(list.Contains(d))
```

Effectively you would make it so that doubles could not actually work effectively as any sort of object in any sort of container.  You could not use them effectively in linq.  They would always be something that never meshed well with the rest of .Net.    So .Net implements .Net semantics.  If you want IEEE semantics, you can use the simple operators.

**Comment by CyrusNajmabadi** [comment]
I'd like the table the Double discussion as the interesting behaviors of Double are not what are important here.  All the reasons for tuples needing == support can be explained with simpler types like int/long.  

For people who don't think == should be lifted to tuples, can i ask why you think the following should behave differently:

```c#
int i = 0;
long j = 0;

Console.WriteLine(i == j); // what do you think this should print?

(int, int) t1 = (i, i);
(long, long) t2 == (j, j);
Console.WriteLine(t1 == t2); // what do you think this shoudl print?
```

If you think the answer should be 'true' then 'false', can you explain why you think that two values which previous were ```== true```, should now not compare the same just because they have been packaged up into a pair of values?

**Comment by DavidArno** [comment]
@CyrusNajmabadi 
````cs
int i = 0;
long j = 0;

Console.WriteLine(i == j); // what do you think this should print?
````
I think it should print exactly the same as for:
````cs
int i = 0;
long j = 0;

Console.WriteLine(i.Equals(j)); 
````
Whether it's `true` or `false` isn't important.

As Eric Lippert says in his [Sharp Regrets: Top 10 Worst C# Features article](http://www.informit.com/articles/article.aspx?p=2425867), equality in C# is one of its top ten worst features. But what is done is done and we must live with it. However, watching you try to defend it, rather than just admitting it was a mistake that we can all learn from, is more than a little frustrating.

I'm still confused as to how you think `==` will work in tuples though, based on your expression:
````
(x1, ..., xn) == (y1, ..., yn)   <==>  x1 == y1 && ... && xn == yn
````
As per my previous example, `struct S {}`, how will the following work?
````
var x = (new S(), new S());
var y = (new S(), new S());
if (x == y) ...
````

**Comment by Thaina** [comment]
@DavidArno What you proposed might be ideal but it could be breaking change so that's a problem. Sometimes we need to deal with legacy

**Comment by CyrusNajmabadi** [comment]
> I think it should print exactly the same as for:

Ok.  Given that that's not how C# has behaved since v1, that seems bad.  You are complaining about inconsistency, but would not like C# vNext to be inconsistent with C# v1-v7.

> As per my previous example, struct S {}, how will the following work?

```c#
var x = (new S(), new S());
var y = (new S(), new S());
if (x == y) ...
```

it's exactly as i stated, it would be equivalent to:

```c#
if (x.Item1 == y.Item1 && x.Item2 == y.Item2)
```

If S has no == operator then i would expect an error.  precisely as i would if i wrote:

```c#
var v1 = new S();
var v2 = new S();
Console.WriteLine(v1 == v2);
```

**Comment by DavidArno** [comment]
@CyrusNajmabadi,

> If S has no == operator then i would expect an error

Good. Right so then on to my next example from above:
````cs
bool F<T1,T2>((T1, T2) x, (T1, T2) y) => x == y;
````
`T1` or `T2` might be `S` here, so there's going to have to be an error somewhere. When will the error occur? At compile time or run time?

**Comment by CyrusNajmabadi** [comment]
> However, watching you try to defend it, rather than just admitting it was a mistake that we can all learn from, is more than a little frustrating.

I don't believe it was a mistake.  

And i'm not even really trying to defend it (not much to defend IMO).  What i'm trying to do is explain to you why it's valuable for == on tuples to retain the semantics that == has today for C#.  Actually breaking that *is* confusing and breaks composition.  Making values composable was a direct goal for tuples, and this really impedes with that.

**Comment by CyrusNajmabadi** [comment]
> When will the error occur? At compile time or run time?

At compile time.  Precisely the same as if you wrote:

```c#
bool F<T1>(T1 x, T1 y) => x == y;
```

You can't write this today in C#, so i don't see why the presence of tuples would change anything.

**Comment by CyrusNajmabadi** [comment]
> I think it should print exactly the same as for:

To be clear, you recommend massively breaking either the majority of C# code out there (if we change the semantics of == to match .Equals for disparate types), or we risk breaking the majority of .net code out there if we change .Equals to match C#'s == 

Neither of these are going to fly.  And insisting that we go with a completely non-viable approach for equality and then that we also make our new features fit this non-existent reality just not going to go anywhere.  Neither .Net or C# are going to break the majority of programs out there.  And as such there are going to be different semantics to == and .Equals. 

The question then becomes "should == on a tuple behave like == on the constituent parts, or like .Equals?"  As you yourself have mentioned many times, the 'principle of least surprise' should be followed.  In this case, that principle would lead to == (for tuples) behaving like == (for elements), just as .Equals (for tuples) behaves like .Equals (for elements).

**Comment by DavidArno** [comment]
@CyrusNajmabadi,
> At compile time. Precisely the same as if you wrote...

Ok. So `bool F<T1>(T1 x, T1 y) => x == y;` gives a compile time error because the compiler knows that `T1` might be a struct and that it therefore may not implement `==`. But this proposal is to implement `==` for value tuples. So they aren't the same at all, let alone precisely the same. 

Now the compiler has to know that, whilst `ValueTuple<T1,T2>` implements `==` (and we'll come to how it's to do that in a second), it must emit an error at compile time if the types of `T1` and `T2` aren't known, or if they are known to be structs that do not implement `==`. That's a big burden you are now laying on the compiler and you are making those tuple types a very special case for the compiler to handle. Is that really something the compiler should be doing?

OK, so we have implemented `==` in these value tuple types. How exactly? You can't implement it in `(T1, T2)` as:
````cs
public static bool operator ==((T1, T2) left, (T1, T2) right) => 
    left.Item1 == right.Item1 && left.Item2 == right.Item2;
````
as `T1` and `T2` aren't constrained to classes, so the above would be a compilation error. But you want `==` in tuples to invoke `==` on its items. How are you going to do that?

**Comment by DavidArno** [comment]
@CyrusNajmabadi & @Thaina,
> To be clear, you recommend massively breaking either the majority of C# code out there

To be very clear: I'm doing no such thing. That's the way it *should* have been implemented. But it wasn't and as I said: "what is done is done and we must live with it". Yes we need to deal with legacy. Thus this whole discussion as there's more than one way of dealing with it, each with their pros and cons. Though I'd argue that just burying our heads and believing no mistake was made, has no pros.

**Comment by CyrusNajmabadi** [comment]
> Now the compiler has to know that, whilst ValueTuple<T1,T2> implements ==

I'm not sure where you're getting that idea from.  There is no way for ValueTuple to effectively implement ==.  To begin with, it wouldn't know how to statically dispatch the appropriate == operator on all the elements.

As i stated earlier, using == on a tuple is *exactly* a transformation of:

```c#
(x1, ..., xn) == (y1, ..., yn)   <==>  x1 == y1 && ... && xn == yn
```

> That's a big burden you are now laying on the compiler

It's not a burden at all.  It is trivial for the compiler to do this check.  If the user writes "someTuple == someOtherTuple" then we will create a node for this.  Then, during typechecking if we see that these are tuples on both sides we'll first check that the arity is the same.  Then we'll go determine if we can find an applicable == operator for each pair of element types between the two tuples.  This is extremely simple to do.

> OK, so we have implemented == in these value tuple types.

== is not implemented in ValueTuple.  That's why this isn't a bug on CoreFx.  This is about C# understanding == and emitting appropriate code for it.  Just as we do already for things like == between a whole host of other types that are special to the language.

> But you want == in tuples to invoke == on its items. How are you going to do that?

I have specified how about 4 times now.  If you write: 

```c#
t1 == t2
```

That that is exactly the same as writing:

```c#
t1.Item1 == t2.Item1 && ... && t1.Item1 == t2.ItemN
```

It is the == operator composed over all the constituent elements of the tuple.

**Comment by CyrusNajmabadi** [comment]
> Thus this whole discussion as there's more than one way of dealing with it, each with their pros and cons.

Can you explain why you think it is a pro for == (for tuples) to behave like .Equals (for tuple elements) instead of like == (for tuple elements)?  It seems to violate the 'principle of least surprise' that you've desired in the past.

Again, we're past the point about what == means for the language.  In C#:

```c#
byte b = 1;
int i = 1;
b == i; // this is true
```

Given that, why is it not surprising to you that the following would be false:

```c#
var t1 = (b, b);
var t2 = (i, i);
t1 == t2;  // this is now false??
```

If 'b' is == to 'i', then why is composing two 'b's together not equal to composing two 'i's together?

**Comment by DavidArno** [comment]
@CyrusNajmabadi,

Ah, my apologies. I misunderstood what you were saying. I hadn't appreciated you were proposing that the compiler transform `(x1, ..., xn) == (y1, ..., yn);` into `x1 == y1 && ... && xn == yn;` and then apply type checking to see if it were allowed. So it will be a compiler trick, not an operator. Makes more sense now.

> Can you explain why you think it is a pro for == (for tuples) to behave like .Equals (for tuple elements) instead of like == (for tuple elements)?

Sure:
1. It would work for all tuples, not just for those with items of types that implement `==`,
2. It would be a simple and consistent approach to teach developers: `==` on tuples uses eg `EqualityComparer `  against each item to determine equality of the tuple. Usual rules of "beware comparisons in C#, for they are screwy" applies.

**Comment by orthoxerox** [comment]
I must say David/Cyrus spats are always the highlight of my day. :grin:

@DavidArno there aren't that many types where `==` and `Equals` produce different results and most of them are numerics, where the difference is more or less justified.

**Comment by DavidArno** [comment]
@orthoxerox,

So it was lucky that Cyrus didn't read your comment and just say "we'll implement it as a compiler trick as orthoxerox suggests" for the entire spat could have been avoided then and you would have missed out on your fun 😆

**Comment by jnm2** [comment]
I don't consider it a compiler trick. The compiler trick is that ValueTuple is involved at all :D tuples != ValueType and I wish people would stop equating the two.

**Comment by CyrusNajmabadi** [comment]
> tuples != ValueTuple

tuples are not necessarily ValueTuples.  But ValueTuples corresponding to a certain shape are always tuples.  As such, any of the == stuff we're talking about would work properly even if you had "ValueTuple<int, string> x, y; ... x == y".

**Comment by jnm2** [comment]
That's a good way to put it. Stealing that later.

**Comment by GeirGrusom** [comment]
@DavidArno

What about reference types? The usual assumption is that `==` for reference types is a reference comparison. Tuples would suddenly break that assumption. I think your proposed behavior could cause quite a lot of late night cursing.

**Comment by DavidArno** [comment]
@GeirGrusom,

Maybe it's just me, but I cannot begin to understand the thought processes of someone who, for reasons other than ignorance, chooses to override `.Equals` on a class, but doesn't override `==` to have exactly the same affect.

**Comment by GeirGrusom** [comment]
@DavidArno that's not what this is about.

**Comment by DavidArno** [comment]
@GeirGrusom,

That's odd, as that is exactly what it is about, as far as I'm aware. What do you feel it's about?

**Comment by GeirGrusom** [comment]
You're applying some standard of behavior among developers and implying that anyone that doesn't share your opinion are poor developers and trying to use that as a point in the discussion. But this isn't about what you think C# developers should be doing. It's about how C# behaves as a language, and suddenly making `==` act like `.Equals` in certain conditions is not consistent behavior. You can't defend an odd behavior change because you feel that developers should write code in a certain way because you would totally do it that way.

**Comment by DavidArno** [comment]
@GeirGrusom,
> You're applying some standard of behavior among developers and implying that anyone that doesn't share your opinion are poor developers and trying to use that as a point in the discussion.

I'm sorry to hear that you have drawn that conclusion from what I've said.

> It's about how C# behaves as a language, and suddenly making `==` act like `.Equals` in certain conditions is not consistent behavior.

There is no consistent behaviour when it comes to C# and equality and we've all had to try and deal with that since v1. As a result, there is no solution that can be adopted for `==` on value tuples that will result in consistency. The team have to choose between a number of solutions that each introduce their own, different, inconsistencies:
1. Do nothing and do not support `==` and `!=` for those tuples, as @HaloFour suggests. This is actually the most consistent of the options as many structs do not implement those operators. I think this is the option I'm starting to favour.
2. Have `==` call `.Equals`. This provides consistency between the two ways of testing for equality in value tuples and it will work for all of those tuples. It's downside is that two tuples may be equal, even if their properties aren't equal if directly tested.
3. Implement a more clever system of testing for `IComparable`, `IEquatable`, determining if structs are involved etc and using different ways of testing the properties of the tuple. This will result in less cases of tuple equality/item inequality than option 2, but will still lead to those situations at times.
4. Implement it as a compiler trick as the team appear to be proposing. This will ensure `==` on a tuple always results in `==` being used for its properties. This will create consistency in that regard. But it'll be a special case type that doesn't use operators for `==`. That special case will only work for some tuples, won't work at all with generics and potentially creates weird effects that'll confuse people when eg tuples of tuples are used.

Option 4 is by far the worst option in my view. It's sad you think me saying so implies I'm judging the abilities of other developers.

**Comment by HaloFour** [comment]
@DavidArno 

I advocate option 4, not option 1.  The elements of the tuple should be directly compared to one another using whatever existing equality semantics the compiler can resolve between the types of their elements.  I believe that is the most intuitive and most internally consistent option (consistent in their inconsistencies).  Tuples aren't proper containers; the compiler should pretend that they don't exist.

The fact that C# and the CLR allows `operator ==` and `Equals` to be different is a boat that has *long* since sailed.  I agree that they generally should be the same, but there are reasons for that to not be the case, such as with `double`.  We can't fix that now, and we shouldn't be trying to drag tuples along for that ride.

**Comment by DavidArno** [comment]
@HaloFour,

My apologies. I could have sworn you'd said as much early in this discussion. But I can't find where I thought you said that, so I'm imagining it. I've removed that false assertion therefore. Sorry.

**Comment by jcouv** [comment]
@DavidArno 
Thanks for summarizing the options.

> But it'll be a special case type that doesn't use operators for ==.

Yes, option 4 introduces special compiler behavior for tuples. We've had to add such special handling for type inference, overload resolution, and many other cases already. 
The guiding principle was to distribute the behavior onto the elements. For instance, picking an overload for `M((1, 2), 3);` should behave like a flattened non-tuple analog `M(1, 2, 3);`. Option 4 keeps that approach, which seems sensible.

I don't expect people to write their own `ValueTuple` types, so the fact that the `==` on tuples doesn't map to the `==` on `ValueTuple` shouldn't really be noticeable. This isn't different to accessing `.Item10`, which doesn't exist in any `ValueTuple`.

>That special case will only work for some tuples

Yes, option 4 would only work on tuples whose elements can be compared with `==`. I'm not sure what that is a negative, as the same rule applies to individual elements (by definition).

> won't work at all with generics 

I think you're referring to the `bool F<T1,T2>((T1, T2) x, (T1, T2) y) => x == y;` example. 
I don't see this example as a downside either, as it behaves the same as the non-tuple case `bool F<T1>(T1 x, T1 y) => x == y;` (as Cyrus pointed out).

> and potentially creates weird effects that'll confuse people when eg tuples of tuples are used.

I looked up the thread and didn't see any examples of nested tuples. Can you provide an example that is confusing?

It seems to me that option 4 works fine with nested tuples. `(a1, (b1, c1)) == (a2, (b2, c2))` will mean the same as `a1 == a2 && (b1, c1) == (b2, c2)` which means the same as `a1 == a2 && b1 == b2 && c1 == c2` (which would be the lowered form).

**Comment by DavidArno** [comment]
@jcouv,

I was over-thinking the nested tuples idea. I went down the `(a1, (b1, c1)) == (a2, (b2, c2))` route by thinking `(b1, c1) == (b2, c2)` couldn't work as then they are value tuples, which do not have the `==` operator. But of course, the compiler can just keep decomposing them, as you say. So scrap the "it'll be confusing" line; that's not true.

**Comment by DavidArno** [comment]
@jcouv,

> I don't expect people to write their own ValueTuple types...

That's a risky assumption to make. I started writing my own version as I wanted ones that were immutable. However I ran into weird compiler errors around `ValueTuple<T1,T2>` not being assignable to `ValueTuple<T1,T2>` and the like. Not really had time to investigate whether that was me doing something odd (though I ruled out having any ref to the official tuples) though.

It might be prudent therefore, if (when) option 4 is implemented to have the compiler warn on any `==` and `!=` operators being added to a custom value tuple implementation that the compiler will ignore them.

**Comment by jnm2** [comment]
> It might be prudent therefore, if (when) option 4 is implemented to have the compiler warn on any == and != operators being added to a custom value tuple implementation that the compiler will ignore them.

As the compiler should. The tuples feature isn't there to serve the ValueTuple imeplementation; The ValueTUple implementation is there to serve the tuples feature. The central philosophy around tuples is that operations on the tuple should be distributive as operations on each individual value.

**Comment by DavidArno** [comment]
@jnm2,

Actually, thinking it through, since the team is adverse to introducing breaking changes, they might need to do the reverse. By the time this feature appears in version 7.X, someone may have written a custom value tuple implementation that implements those operators. Overriding them could then be a breaking change. So the compiler might need to respect those operators if present, and use the built-in rules if not.

**Comment by CyrusNajmabadi** [comment]
I'm with @jnm2 on this.  ValueTuple is simply there to serve the language.    I'm fine with this being totally at a language level, and ValueTuple not being observed at all once we do this work.

**Comment by HaloFour** [comment]
@DavidArno 

> Actually, thinking it through, since the team is adverse to introducing breaking changes, they might need to do the reverse. By the time this feature appears in version 7.X, someone may have written a custom value tuple implementation that implements those operators. Overriding them could then be a breaking change. So the compiler might need to respect those operators if present, and use the built-in rules if not.

Conceptually I agree, however it seems that the team has already made post-C# 7.0 decisions where the compiler may skip the container and work with the elements directly:

https://github.com/dotnet/roslyn/issues/16869

**Comment by DavidArno** [comment]
@HaloFour,

Ha ha, hoisted by my own petard, there! Good call. 🏁

**Comment by alrz** [comment]
I wonder if this could be implemented [on top of shapes](https://doc.rust-lang.org/std/cmp/trait.Eq.html) rather than hard-coding to the compiler.

**Comment by jcouv** [comment]
@DavidArno  I'd like to add one clarification (related dotnet/roslyn#16869): both C# and VB languages and compilers now assume that the `ValueTuple` constructor is side-effect-free, so it can be skipped in some cases.
Out of curiosity, would you mind sharing some scenarios that motivate you to implement your own `ValueTuple` type?

Thanks @HaloFour for making the connection between those two discussions.

**Comment by DavidArno** [comment]
@jcouv,

I have (well had, see below) a single motivation. I don't like the fact that value tuples are mutable. The natural conclusion (in my head, anyway) is therefore to create my own - immutable - version.

However, I've reconsidered that plan, based on two things:
1. The compiler will, from 7.1 onwards (or is it in 7?) make assumptions about the behaviour of those tuples,
2. I read today (from a Roslyn issue you were involved in, I think, but I can't find it now), that from eg .NET framework v4.7, the ValueTuple and new async feature nuget packages will be baked into the framework.

Therefore, I no longer consider it wise to create a custom ValueTuple implementation.

**Comment by HaloFour** [comment]
@alrz

> I wonder if this could be implemented on top of shapes rather than hard-coding into the compiler. Not only equality, almost all of tuple operations should be distributed over all elements (if all of them satisfy a certain set of constraints which is representable through generic shapes).

e.g. `using` over a tuple full of `IDisposable`s or `await` over a tuple full of elements for which the compiler can resolve `GetAwaiter()`?

In those cases it might make sense, but I think it might not always be so obvious how an operation should be distributed, if it should be at all.  For the most appropriate behavior it might be necessary for the compiler to special-case the operations in question.

**Comment by alrz** [comment]
> I think it might not always be so obvious **how an operation should be distributed**, if it should be at all. For the most appropriate behavior **it might be necessary for the compiler to special-case the operations** in question.

That directly depends on the shape implementation. I'm not sure why "special-casing" would make that more "appropriate". I think this would be the most sensible starting point for designing shapes. Once it's possible to model various operations through shapes there is no reason to special-case such features. IMO it'd be worth it to invest on the general solution rather than special-casing all these into the compiler.

**Comment by Richiban** [comment]
@alrz 

> Once it's possible to model various operations through shapes there is no reason to special-case such features

I would normally agree with you, except I do think that it might be appropriate in the case of tuples. Because they should really be such a basic and integral feature I'm all in favour of the compiler 'inlining' the body of the '==' method for tuples, even if the method exists (at least notionally) as the operator '==' on a ValueTuple where T1, T2 etc are members of an Eq typeclass.

**Comment by alrz** [comment]
The case for inlininig was discussed at length in other threads before and concluded that roslyn is not the place for these optimizations (even though there were some good arguments against it). In fact, that was my first concern regarding how shapes are implemented (https://github.com/dotnet/csharplang/issues/164) but as others have said, invocations of shape methods get specialized by the JIT.

**Comment by jcouv** [comment]
@alrz This feature allows comparing two tuple variables (`tuple1 == tuple2`) as well, whereas positional patterns require constants.

**Comment by DavidArno** [comment]
@jcouv,

That doesn't sound right to me. Surely positional patterns for tuples will support recursive patterns, not just constants?
```
tuple is (var x, (1, var y),  3)
```

@alrz,
Surely "I think it's not, other than that you could compare variables with `==`, right?" applies to all basic patterns, doesn't it?
```
x is 1
x == 1
```

**Comment by HaloFour** [comment]
@DavidArno 

Yes, but the only actual values that you can compare must be themselves constants.  So if you're comparing to other variables via pattern matching you'd have to use pattern variables and guards:

```cs
public void Foo((int, int) x, (int, int) y) {
    switch (x) {
        case (var a, var b) when a == y.Item1 && b == y.Item2: ...
    }
    // vs
    if (x == y) { ... }
}
```

I assume that you could also use it as shorthand for comparing equality of multiple items:

```cs
public void Foo(int a, int b, int c, int d, int e, int f) {
    if (a == d && b == e && c == f) { ... }
    // vs
    if ((a, b, c) == (d, e, f)) { ... }
}
```

**Comment by alrz** [comment]
> values that you can compare must be themselves constants.

I think this restriction can be relaxed at least for `is` (not for `case` cause it needs to check for subsumption). that would make the two identical.

**Comment by DavidArno** [comment]
@HaloFour 

I love the idea of:
```cs
public void Foo(int a, int b, int c, int d, int e, int f) {
    if ((a, b, c) == (d, e, f)) { ... }
}
```
Hopefully, @jcouv can work the same magic as he did for the immodestly named [Arno Assignment Pattern](https://github.com/dotnet/roslyn/issues/16869) so that the tuples get optimised away here too.

**Comment by jcouv** [comment]
@DavidArno `(a, b, c) == (d, e, f)` will likely require referencing the `ValueTuple` types for the compilation to succeed, but no `ValueTuple` reference will be emitted in IL.

**Comment by jnm2** [comment]
@jcouv Couldn't the ValueTuple requirement be fixed the same way it was for deconstruction?

**Comment by jcouv** [comment]
@jnm2 It's not obvious. Still looking into it ;-)
